### ISSUE TITLE
deliveryChanges does not work with Dart2 (with solution)

### ISSUE THREAD

**Issue by jodinathan** [dart-lang/observable#54]
Extend *ChangeNotifier* with your own *ChangeRecord*, like:
``` 
class MyRecord extends ChangeRecord {
}
class MyClass extends ChangeNotifier<MyRecord> {
}
```

once you add some notifications, dart throws an error on:

```
bool deliverChanges() {
    List<ChangeRecord> changes;
    ...
    _changes.add(changes); // this will throw an error of mismatched types
}
```

The only thing you need to do to solve it is change this line:

```
bool deliverChanges() {
    List<ChangeRecord> changes;
```

to this:

```
bool deliverChanges() {
    List<C> changes;
```

**Comment by natebosch** [comment]
I assume you're seeing the issue at runtime in DDC? what version of the SDK are you on?

**Comment by jodinathan** [comment]
sorry.
DDC
dev-24 and dev-28 neither worked

**Comment by natebosch** [comment]
If we change the type to `List<C>` then I think the assignment to `ChangeRecord.ANY` would fail.

https://github.com/dart-lang/observable/blob/bc4b2d9902099e459ea744b8969239156c65623f/lib/src/change_notifier.dart#L61

`ChangeRecord.ANY` is a `List<ChangeRecord>` which is not a `List<C>`

https://github.com/dart-lang/observable/blob/bc4b2d9902099e459ea744b8969239156c65623f/lib/src/records.dart#L22

I think this is likely to be a bit tougher to solve and I'm not sure if we have the bandwidth to dig deeply on it...

**Comment by jodinathan** [comment]
yeah, just tested it and it threw an error with ANY.
that is probably why it was ```List<ChangeRecord>``` in the first place.

Maybe just sending an empty instead of the ANY const?
afterall, that dummy ChangeRecord wasn't much of a help.

Or you can store the last queue and send it again if the new queue is empty.

**Comment by kevmoo** [comment]
CC @nshahan

**Comment by jodinathan** [comment]
For now I am using my fork: https://github.com/jodinathan/observable
It basically changes:

https://github.com/dart-lang/observable/blob/bc4b2d9902099e459ea744b8969239156c65623f/lib/src/change_notifier.dart#L55

to:
`List<C> changes;`

and

https://github.com/dart-lang/observable/blob/bc4b2d9902099e459ea744b8969239156c65623f/lib/src/change_notifier.dart#L61

to:

`changes = <C>[];`

It is a breaking change, however, that empty ChangeRecord from ANY const has no value at all.

**Comment by shamblett** [comment]
Hi guys, any update on this as far as a permanent fix is concerned, I'd rather use observable if a fix is forthcoming.

**Comment by kevmoo** [comment]
@nshahan ?

**Comment by nshahan** [comment]
I'll do some testing on our codebase to see if we can easily fix up any breaks.

**Comment by buhhy** [comment]
I've created a pull request to fix this issue:

https://github.com/dart-lang/observable/pull/76

**Comment by wsakka** [comment]
Not sure why the flutter team leaves these issues open -- it's been a year, what's the status of this issue?

**Comment by jodinathan** [comment]
@wsakka it has been fixed with version 0.22.1+5. 
I am closing this issue.
